### ISSUE TITLE
Server crash after transforming equipped item?

### ISSUE THREAD

**Issue by nekiro** [otland/forgottenserver#2280]
### Before creating an issue, please ensure:
- [x] This is a bug in the software that resides in this repository, and not a
      support matter (use https://otland.net/forums/support.16/ for support)
- [x] This issue is reproducible without changes to the code in this repository

### Steps to reproduce (include any configuration/script required to reproduce)
1. make onequip movement with this code:
```
function onEquip(player, item, slot)
	local slotItem = player:getSlotItem(CONST_SLOT_RING)
	if not slotItem or slotItem.itemid ~= 22543 then
		return false
	end
	
	slotItem:transform(22516)
	slotItem:setAttribute(ITEM_ATTRIBUTE_DESCRIPTION, 'This constantly rotating ring is perfectly weightless, yet feels infinitely heavy at the same time.')

	return true
end
```

I discovered this accidentally.

**Comment by XKaoo** [comment]
I'm very new here (and to Tibia in general) but I tried to reproduce it without any success (as in, it didn't make the server crash).
What I did:  
- Set my account type to 5 and group ID to 3
- Added this line to `talkactions.xml` under `test talkactions`:
`<talkaction words="!healthtest" separator=" " script="healthtest.lua" />`
- Put this in `healthtest.lua`:
```
function onSay(player, words, param)
	player:addHealth(-player:getHealth())
	player:sendTextMessage(MESSAGE_STATUS_CONSOLE_BLUE, "executed test talkaction !healthtest")
end
```  
- Reloaded talk actions with the `reload` command

I did see the "executed test talkaction" message as well as the "smoke" animation on the character but the server didn't crash at all. I've also tried to restart it and type the command again but no crash either.

**Comment by thexamx** [comment]
Hmmm, I was it like this in onequip movement, maybe thats the case?

**Comment by ranisalt** [comment]
Just in case, run it on GDB to check exactly which function crashes.

**Comment by thexamx** [comment]
https://pastebin.com/raw/Fh4EDe6x

Full crash log

**Comment by 4drik** [comment]
sill crashing?
```
function onEquip(player, item, slot)
	local slotItem = player:getSlotItem(CONST_SLOT_RING)
	if not slotItem or slotItem and slotItem.itemid ~= 22543 then
		return false
	end
	
	slotItem:transform(22516)

	return true
end
```

**Comment by ranisalt** [comment]
Looks like a double free:

```#3  0x00007ffff602482e in malloc_printerr (ptr=<optimized out>, 
    str=0x7ffff612645e "free(): invalid pointer", action=1) at malloc.c:4998
        buf = "00007ffff766f0d0"
        cp = <optimized out>```

**Comment by thexamx** [comment]
@4drik yes.
@ranisalt 
I dont really understand crash logs, but I workaround this with slotItem:remove() and then adding new ring to the slot, but unfortunately is crashing my client with "index out of range" debug window.

**Comment by ranisalt** [comment]
I think internally TFS replaces the item with a new copy, throws the old one out, and somewhere after the function it tries to discard it again, this time a null pointer. Gonna have a look later.

**Comment by Codex-NG17** [comment]
I've actually encountered this but I found a solution, I have to look for the code. It's written in Lua not C++.

**Comment by Codex-NG17** [comment]
This is part of my magical items library, I haven't had any issues with transformations since then.
https://github.com/Codex-NG17/Magical-Items-1.2-3-Otx3/blob/master/magical_items_lib.lua#L1221-L1249

```
function Player:transformMagicItemTo(item, slot, name)
    if x.transform[slot] and next(x.transform[slot]) then
        local myItem = x.transform[slot][name]
        if myItem and next(myItem) then
            for _, tbl in pairs(myItem) do
                for property, value in pairs(tbl) do
                    if property == 'on' then
                        item:transform(value)
                    end
                end
            end
        end
    end
end

function Player:transformMagicItemFrom(item, slot, name) 
    if x.transform[slot] and next(x.transform[slot]) then
        local myItem = x.transform[slot][name]
        if myItem and next(myItem) then
            for _, tbl in pairs(myItem) do
                for property, value in pairs(tbl) do
                    if property == 'off' then
                        item:transform(value)
                    end
                end
            end
        end
    end
end
```
I know I should just consolidate the code into one method & optimize the rest of the library, but I am working in another framework which isn't tibia or ot related.

**Comment by nekiro** [comment]
I gave this issue a try once again and I found out new informations.
This code I provided does not crash the server. The server crashes code which is set to onDeEquip to transform back item to whatever, it happens, because when we transform item we create new and we don't discard old (is still kept in memory, but as removed) and onDeEquip still holds userdata of the old item which does not exist now, and that's what crashes the server, because it tries to deequip "deleted" item.

`function onEquip(player, item, slot)
	addEvent(function() 
		local slotItem = player:getSlotItem(CONST_SLOT_RING)
		print(slotItem, slotItem:getType():getName())
		slotItem:transform(11134)
	end, 5 * 1000)
	print("wtf")
	return true
end

function onDeEquip(player, item, slot)
	print(item, item:getName())
	item:transform(13877)
	return true
end`
